clang/lib/Sema/SemaType.cpp

```diff
@@ -2322,7 +2322,7 @@ QualType Sema::BuildArrayType(QualType T, ArrayType::ArraySizeModifier ASM,
   // OpenCL v2.0 s6.12.5 - Arrays of blocks are not supported.
   // OpenCL v2.0 s6.16.13.1 - Arrays of pipe type are not supported.
   // OpenCL v2.0 s6.9.b - Arrays of image/sampler type are not supported.
-  if (getLangOpts().OpenCL) {
+  if (getLangOpts().OpenCL || getLangOpts().SYCLIsDevice) {
```

Comment:
Following the logic: 
> Sampler type is enabled only for OpenCL and SYCL modes, so if R->isSamplerT() returns true, we know that getLangOpts().SYCLIsDevice || getLangOpts().OpenCL is also true.

Can't we remove getLangOpts checks?

Reply:
I don't think so.
Blocks are not OpenCL specific types.